### PR TITLE
Check for error

### LINKED ISSUE #7051: Azure connections go in a loop to open connection dialog after connection failed + cancel
- Azure Data Studio Version: (release branch)
Version: 1.11.0 (user setup)
Commit: acc85c2c23f4889b4ae9373282d92daec321b1bd
Date: 2019-09-03T23:55:40.484Z
VS Code 1.37.0
Electron: 4.2.9
Chrome: 69.0.3497.128
Node.js: 10.11.0
V8: 6.9.427.31-electron.0
OS: Windows_NT x64 10.0.17763

Steps to Reproduce:

1. Connect to an Azure account.
2. expand to subscription-->Servers 
3. Click on a server that you don't have login creds for
4. you get a connection failed pop up (as expected) and then you cancel the connection dialog
5. After this there is a loop of error popup and connection dialog reopening like following
6. To break the loop is to manually enter some valid connection details (for any other server) and click connect
7. [updated] But step 6 causes this server with no access to show some random children (Tables views etc without the database level). I am not sure if it is showing the actual data or from some default place because all of them seem to show same nodes

loop in azure

image

Error log from connection error : 

System.Data.SqlClient.SqlException (0x80131904): Login failed for user <username>.
   at System.Data.SqlClient.SqlInternalConnectionTds..ctor(DbConnectionPoolIdentity identity, SqlConnectionString connectionOptions, SqlCredential credential, Object providerInfo, String newPassword, SecureString newSecurePassword, Boolean redirectedUserInstance, SqlConnectionString userConnectionOptions, SessionData reconnectSessionData, Boolean applyTransientFaultHandling, String accessToken)
   at System.Data.SqlClient.SqlConnectionFactory.CreateConnection(DbConnectionOptions options, DbConnectionPoolKey poolKey, Object poolGroupProviderInfo, DbConnectionPool pool, DbConnection owningConnection, DbConnectionOptions userOptions)
   at System.Data.ProviderBase.DbConnectionFactory.CreateNonPooledConnection(DbConnection owningConnection, DbConnectionPoolGroup poolGroup, DbConnectionOptions userOptions)
   at System.Data.ProviderBase.DbConnectionFactory.<>c__DisplayClass40_0.<TryGetConnection>b__1(Task`1 _)
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.SqlTools.ServiceLayer.Connection.ReliableConnection.ReliableSqlConnection.<>c__DisplayClass28_0.<<OpenAsync>b__0>d.MoveNext() in D:\a\1\s\src\Microsoft.SqlTools.ManagedBatchParser\ReliableConnection\ReliableSqlConnection.cs:line 303
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.SqlTools.ServiceLayer.Connection.ConnectionService.TryOpenConnection(ConnectionInfo connectionInfo, ConnectParams connectionParams) in D:\a\1\s\src\Microsoft.SqlTools.ServiceLayer\Connection\ConnectionService.cs:line 521
ClientConnectionId:90062522-1f2b-4df1-a269-4d4bf7fc6d8a
Error Number:18456,State:1,Class:14

### PR DESCRIPTION
**Pull request by aaomidi**
Fixes #7051 

I really don't like this solution, so let me know if anyone has a better idea.

### CODE REVIEW

File: src/sql/workbench/browser/parts/views/customView.ts [hunk 1]
```diff
@@ -700,6 +700,9 @@ class TreeDataSource implements IAsyncDataSource<ITreeItem, ITreeItem> {
 				// So in order to enable this we need to tell the tree to refresh this node so it will ask us for the data again
 				setTimeout(() => {
 					this.treeView.collapse(node);
+					if (e.message === 'User canceled') {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/browser/parts/views/customView.ts]: A better solution I'd say would be to have the error thrown at https://github.com/microsoft/azuredatastudio/blob/ea0f9e6ce95b45def2b694efcd8ffe909bc43d20/src/sql/workbench/parts/objectExplorer/browser/objectExplorerViewTreeShim.ts#L93 be a custom error class instead and check for that.

It seems like having cancel be an error at all is kind of weird (I'd think just returning undefined would be fine), but that's not something I think we should be changing at this time so doing a quick check here is a fine enough solution for now I'd say.

Comment (aaomidi) [reply src/sql/workbench/browser/parts/views/customView.ts]: Definitely a better solution here, thanks @Charles-Gagnon.

**Comment by aaomidi** [comment]
Because typescript is compile time only we lose the type of the error in runtime which means I still need to do a string comparison, but at least it's more controlled now.

File: src/sql/workbench/browser/parts/views/customView.ts [hunk 2]
```diff
@@ -700,6 +701,9 @@ class TreeDataSource implements IAsyncDataSource<ITreeItem, ITreeItem> {
 				// So in order to enable this we need to tell the tree to refresh this node so it will ask us for the data again
 				setTimeout(() => {
 					this.treeView.collapse(node);
+					if (e.code === UserCancelledConnectionError.code()) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/browser/parts/views/customView.ts]: You should be able to do

`e instanceof UserCancelledConnectionError`

If UserCancelledConnectionError was an interface that is indeed lost, but actual classes are supported in Javascript so you're able to do the checks just fine https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes

Comment (aaomidi) [reply src/sql/workbench/browser/parts/views/customView.ts]: Yeah you're right. For some reason I thought I was making an interface.

File: src/sql/base/common/errors.ts [hunk 3]
```diff
@@ -10,3 +10,10 @@ export function invalidProvider(name?: string): Error {
 		return new Error('Invalid provider');
 	}
 }
+
+export class UserCancelledConnectionError extends Error {
+	public readonly code: string;
```
Comment (Charles-Gagnon) [inline src/sql/base/common/errors.ts]: Can get rid of code var

File: src/sql/base/common/errors.ts [hunk 4]
```diff
@@ -10,3 +10,10 @@ export function invalidProvider(name?: string): Error {
...
+	constructor(message?) {
```
Comment (Charles-Gagnon) [inline src/sql/base/common/errors.ts]: nit: type here

**Comment by aaomidi** [comment]
This created another issue. You can't click on this node to connect again.

This seems to be an issue that goes all the way up to vscode, so I'll do some more investigations on this.

**Pull request merged**